proposals/readonly-methods.md

```diff
@@ -0,0 +1,65 @@
+# Readonly Methods
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+## Summary
+[summary]: #summary
+
+Provide a way to specify individual methods on a struct do not modify state, in the same way that `readonly struct` specifies no instance method modifies state.
+
+It is worth noting that `readonly method` != `pure method`. A `pure` method guarantees no state will be modified. A `readonly` method only guarantees that instance state will not be modified.
+
+All instance methods on a `readonly struct` could be considered implicitly `readonly methods`. Explicit `readonly methods` declared on non-readonly structs would behave in the same manner. For example, they would still create hidden copies if you called an instance method (on the current instance or on a field of the instance) which was itself not-readonly.
```

Comment:
> For example, they would still create hidden copies if you called an instance method (on the current instance or on a field of the instance) which was itself not-readonly.

I feel like this will be surprising to people.  It think a lot of the motivation here is that "oh, if i have readonly methods, i can help avoid copies". When doing simple stuff causes copying, that seems like it woudl go against that intent.  Woudl it perhaps be better to instead have this error thus making it clear: these methods will will not copy.  If users hit this and then they want to actually perform that functionality that would copy, they can then realize the method will not be readonly, or, they can figure out what they need to update downstream to enable things to stay readonly.

If users really are ok with the copy, i would say that they should then write the copy explicitly themselves.  i.e.:

```c#
var copy = this;
copy.NonReadOnlyMethod();
```

This will make the code nice and explicit and make it much more clear waht it going on.

Reply:
@CyrusNajmabadi, yes that would be ideal. Unfortunately, we already shipped `readonly structs` this way, and making `readonly method` behave differently may be more confusing or more error prone.

I believe the current answer is to provide an analyzer (or warning wave support) to help flag these. CC. @jaredpar

Reply:
That seems reasonable to me.  I would suggest whoever does this work also does hte work to make such an analyzer.  that was we know we have enough information in the compilation API to make it clear when copies are happening :)

Reply:
The analyzers to detect silent struct copies already exist 😄 

https://www.nuget.org/packages/ErrorProne.NET.Structs/0.1.0
https://github.com/SergeyTeplyakov/ErrorProne.NET

CC @SergeyTeplyako